[#1096] Provide the AbstractFoldingTest abstract base class.

miklossy opened the pull request:
- Extend the 'Xtext UI Testing' test infrastructure by the
AbstractFoldingTest abstract class to provide a base infrastructure
for testing the folding capabilities.
- Extend the Xtext Examples the FoldingTest test cases to
demonstrate the usage of the AbstractFoldingTest class.

Signed-off-by: Tamas Miklossy <miklossy@itemis.de>

szarnekow commented:
It would be great if we could test LSP based folding and Eclipse folding with the same test suite.

szarnekow commented on org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui.tests/src/org/eclipse/xtext/example/fowlerdsl/ui/tests/StatemachineFoldingTest.xtend:
```diff
@@ -0,0 +1,92 @@
+				panelClosed => idle
+			end
+		'''.hasFoldingRegions(
+			21 -> 24,
```
I find this very hard to read. Could we annotate the DSL text to mark folding regions?

miklossy replied:
Actually that was also my first idea to annotate the DSL text with the `«c»` special symbols (similar to the other ui test cases) to mark the start end the end positions of the expected folding regions. However, that works only if the expected folding regions are disjunct regions. Considering the corner cases (e.g: overlapping regions, duplicated regions, ...) the annotated DSL would become much more difficult to read.

cdietrich commented:
having a folding test for lsp and ui would require
- having a non eclipse/lsp api (contribution from @tivervac still outstanding)
- having an adapter from non ui to ui.

=> i am not sure if this can be done now, but introducing a new api for eclipse only is bad too

miklossy commented:
As discussed with @cdietrich , I annotated the AbstractFoldingTest API class as Beta.

cdietrich commented:
@szarnekow what do you think?

szarnekow commented:
Unfortunately I'm still not convinced by the internal DSL `hasFoldingRegions(n->m)`

miklossy commented:
What about prefixing the input dsl text with the line numbers?
```
	@Test def complex() {
		'''
		001	events
		002		doorClosed   D1CL
		003		drawerOpened D2OP
		004		lightOn      L1ON
		005		doorOpened   D1OP
		006		panelClosed  PNCL
		007	end
		008	
		009	resetEvents
		010		doorOpened
		011		doorClosed
		012	end
		013	
		014	commands
		015		unlockPanel PNUL
		016		lockPanel   NLK
		017		lockDoor    D1LK
		018		unlockDoor  D1UL
		019	end
		020	
		021	state idle
		022		actions {unlockDoor lockPanel}
		023		doorClosed => active
		024	end
		025	
		026	state active
		027		drawerOpened => waitingForLight
		028		lightOn      => waitingForDrawer
		029	end
		030
		031	state waitingForLight
		032		lightOn => unlockedPanel
		033	end
		034	
		035	state waitingForDrawer
		036		drawerOpened => unlockedPanel
		037	end
		038	
		039	state unlockedPanel
		040		actions {unlockPanel lockDoor}
		041		panelClosed => idle
		042	end
		'''.hasFoldingRegions(
			21 -> 24,
			26 -> 29,
			31 -> 33,
			35 -> 37,
			39 -> 42
		)
	}
```

szarnekow commented:
I'm thinking in the direction of 
```
assertFoldingRegions('''
[>[>events
  doorClosed   D1CL
  drawerOpened D2OP
  lightOn      L1ON
  doorOpened   D1OP
  panelClosed  PNCL
end<]

[>resetEvents
  doorOpened
  doorClosed
end<]<]
''')
```
with an overload that accepts to delimiters.

miklossy commented:
I like the idea of @szarnekow . I updated the PR accordingly.

cdietrich commented:
@szarnekow please examine again

szarnekow commented on org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.xtend:
```diff
@@ -0,0 +1,115 @@
+	@Inject protected extension IFoldingRegionProvider
+
+	/**
+	 * Speical symbol indicating the beginning of a folding region.
```
Minor: Typo

szarnekow commented on org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.xtend:
```diff
@@ -0,0 +1,115 @@
+	/**
+	 * Test that the expected folding regions are present on a given DSL text.
+	 *
+	 * @param it The DSL text where the expected folding regions are enclosed by the {@link #b} and {@link #e} symbols.
```
Is this the correct JavaDoc format for method references?

szarnekow commented on org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.xtend:
```diff
@@ -0,0 +1,115 @@
+	def String b() '''[>'''
+
+	/**
+	 * Speical symbol indicating the end of a folding region.
```
Minor: Typo

szarnekow commented on org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui.tests/src/org/eclipse/xtext/example/domainmodel/ui/tests/FoldingTest.xtend:
```diff
@@ -0,0 +1,131 @@
+			[>entity HasAuthor {
+				author: String
+				
+			[>[>op getAuthor() {
```
Looks like we do create duplicate folding regions in our default impl?

miklossy commented:
Thanks for the review @szarnekow . I fixed the typos and corrected the method references in the javadoc.

Your remark regarding the duplicate folding regions is correct, on the IFoldingRegionProvider level it is possible to create duplicated regions. However, these regions will be filter out in the DefaultFoldingStructureProvider level. See also [PR](https://github.com/eclipse/xtext-eclipse/pull/944) that has already been merged.

szarnekow commented on org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.xtend:
```diff
@@ -0,0 +1,115 @@
+	/**
+	 * Test that the expected folding regions are present on a given DSL text.
+	 *
+	 * @param it The DSL text where the expected folding regions are enclosed by the {@link #b b} and {@link #e e} symbols.
```
Please fix the method references in the JavaDoc here, too (should be `#b()`)

miklossy replied:
done
